airbyte-connector-builder-server/connector_builder/impl/default_api.py

```diff
@@ -2,21 +2,154 @@
+                else:
+                    single_slice.pages.append(message_group)
+        except Exception as error:
+            # TODO: We're temporarily using FastAPI's default exception model. Ideally we should use exceptions defined in the OpenAPI spec
+            raise HTTPException(status_code=400, detail=f"Could not perform read with with error: {error.args[0]}")
+
+        return StreamRead(logs=log_messages, slices=[single_slice])
+
+    def _get_message_groups(self, messages: Iterable[AirbyteMessage]) -> Iterable[Union[StreamReadPages, AirbyteLogMessage]]:
+        """
+        Message groups are partitioned according to when request log messages are received. Subsequent response log messages
+        and record messages belong to the prior request log message and when we encounter another request, append the latest
+        message group.
+
+        Messages received from the CDK read operation will always arrive in the following order:
+        {type: LOG, log: {message: "request: ..."}}
+        {type: LOG, log: {message: "response: ..."}}
+        ... 0 or more record messages
+        {type: RECORD, record: {data: ...}}
+        {type: RECORD, record: {data: ...}}
+        Repeats for each request/response made
+
+        Note: The exception is that normal log messages can be received at any time which are not incorporated into grouping
+        """
+        first_page = True
+        current_records = []
+        current_page_request: Optional[HttpRequest] = None
+        current_page_response: Optional[HttpResponse] = None
+        for message in messages:
+            if first_page and message.type == Type.LOG and message.log.message.startswith("request:"):
+                first_page = False
+                request = self._create_request_from_log_message(message.log)
+                current_page_request = request
+            elif message.type == Type.LOG and message.log.message.startswith("request:"):
+                if not current_page_request or not current_page_response:
+                    self.logger.warning("Every message grouping should have at least one request and response")
+                yield StreamReadPages(request=current_page_request, response=current_page_response, records=current_records)
+                current_page_request = self._create_request_from_log_message(message.log)
+                current_records = []
+            elif message.type == Type.LOG and message.log.message.startswith("response:"):
+                current_page_response = self._create_response_from_log_message(message.log)
+            elif message.type == Type.LOG:
+                yield message.log
+            elif message.type == Type.RECORD:
+                current_records.append(message.record.data)
+        else:
+            if not current_page_request or not current_page_response:
+                self.logger.warning("Every message grouping should have at least one request and response")
```

Comment:
I think that's a fair point i'll change this to throw an error instead of log. The model does specify them as optional, but it makes sense to be more zealous about throwing errors for now since it'll be a louder signal as we roll this out and we can always pull them back as necessary.